### PR TITLE
[Core] Add Repeat Key ("repeat last key") as a core feature.

### PR DESCRIPTION
**Pull request by getreuer**
This PR implements Repeat Key as a core feature. I previously implemented a [userspace Repeat Key implementation](https://getreuer.info/posts/keyboards/repeat-key/index.html) that this one is based on.

## Description

The Repeat Key performs the action of the last pressed key. There are two ergonomically-grounded motivations for Repeat Key:

* [As pointed out by NotGate](https://notgate.github.io/layout/), it is attractive from the point of view that double-tapped letters are like same-finger bigrams (SFBs): "The repeat key fits the theme of reducing SFBs by as much as possible and has the bonus of including the 10th finger. Words like follow become 'fol[rep]ow' so your flow is never broken by the speed of a single finger."

* It is useful for hotkey chords, like repeating Ctrl+Shift+Right to select by word or repeating Ctrl+PageUp / Ctrl+PageDown to switch browser tabs. 

On that second point, hotkeys are quite often in pairs, navigating or acting in one direction or the reverse. This feature also includes a Reverse Repeat Key for such situations. The Reverse Repeat Key performs the "reverse" of the last key, if a reverse is defined. When the last key wast Ctrl+Shift+Right, the Reverse Repeat Key performs Ctrl+Shift+Left. Together with the Repeat Key, this enables convenient selection by words in either direction.

The implementation interoperates well with other features. It is essentially a one-key version of how dynamic macros work: pressing the Repeat Key (or Reverse Repeat Key) generates a `keyrecord_t` and passes it to `process_record()`. This interoperates predictably with all QMK features that I have tested it with. This includes, for example, macros with custom handlers in `process_record_user()`: tapping the macro key and then Repeat Key invokes the macro twice. 

The implementation is space efficient. On my keymap, it adds 686 bytes to the firmware size. If Reverse Repeat Key is disabled (there is a `NO_REVERSE_REPEAT_KEY` option), the cost reduces to 350 bytes.

Please see the documentation `docs/feature_repeat_key.md` for further description.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/process_keycode/process_repeat_key.h
Comment (zvecr) [inline quantum/process_keycode/process_repeat_key.h:18]: Including itself?

Comment (getreuer) [reply quantum/process_keycode/process_repeat_key.h]: Whoops, thanks for pointing that out! I removed it.

**Comment by getreuer** [comment]
Here is an outline for navigating the code changes in this PR.

## Setup

* I defined two new keycodes `QK_REP` and `QK_RREP` in
  keycodes_0.0.2_quantum.hjson and ran `qmk generate-keycodes -v latest`. 

* Repeat Key leverages the `.keycode` field in `keyrecord_t`.
  This field currently exists only when Combos are enabled, so I modified
  action.h and action.c where needed to use this field when either Combos or
  Repeat Key (or both features) are enabled. 

## Sequence of events for Repeat Key

1. On every key press, `process_repeat_key()` calls
   `get_repeat_key_eligible()` to check whether the key is "eligible"
   for repeating. 

2. If so, its keycode and keyrecord as well as the current mods
   state are stored. The keycode is saved in the `.keycode` field (this matters
   if the layer state changes between now and when Repeat is used).

3. When Repeat is pressed (or released), `repeat_key_invoke()` is called.

4. `repeat_key_invoke()` creates a keyrecord by combining the stored keyrecord
   with the current keyevent so that the time and keypos correspond to when 
   Repeat was used.

5. The keyrecord is passed to `process_record()` to execute its effect.

## Reverse Repeat

Reverse Repeat works the same way except for the keycode. When Reverse
Repeat is pressed, the keyrecord's `.keycode` field is set to
`get_rev_repeat_key_keycode()`, getting the "reverse" of the last key.

**Comment by drashna** [comment]
`qmk generate-keycodes-tests -v latest` should be ran too.

**Comment by getreuer** [comment]
> `qmk generate-keycodes-tests -v latest` should be ran too.

Thanks, @drashna! I updated test_common/keycode_table.cpp.

FWIW, I just learned there is a handy script [util/regen.sh](https://github.com/qmk/qmk_firmware/blob/develop/util/regen.sh) that runs all keycode regenerators.

**Comment by drashna** [comment]
Ah, the test function PR was merged.  I wasn't sure if it was or not yet. 

But yeah, that works for that. 

That said, there are some merge conflicts here, as well as keycode value collisions due to the tri layer PR getting merged first.

**Comment by getreuer** [comment]
Done, merged with the develop branch. It's interesting to see the tri layer changes.

**Comment by Ikcelaks** [comment]
I am currently running this branch and using both the repeat key and reverse repeat key in my layout. Thus far both have worked exactly as I expected them to.

I have no comments on the code, but I do question the naming of the Reverse Repeat functionality. This key can be fully dynamic based on the last key press and need not be constrained to logically reversing that keypress. In my keymap, I think of it as the "magic" key, and have built a dictionary that eliminates the worst single finger bigrams that would otherwise exist on the layout. I suggest that the keycode be something like `QK_DYNAMIC_REPEAT` to reflect that it functions like the repeat key, except that the output is fully dynamic based on your definition of `get_rev_repeat_key_keycode_user()`.

I think it would be worth adding some documentation that highlights this use-case, even if it is still called Reverse Repeat.

**Comment by getreuer** [comment]
@Ikcelaks thanks for checking out the PR and this excellent point! You are right, Reverse Repeat can be used for much more than "reversing", and this is well worth clarifying. Leveraging it to remove SFBs is an awesome idea, you inspire me to experiment with that.

I renamed Reverse Repeat to Alternate Repeat (and "`rev_repeat_key`" APIs renamed to "`alt_repeat_key`) to suggest a wider scope including the original "reverse" use case as well as dynamic or complementary actions more generally.

**Comment by Ikcelaks** [comment]
`alt_repeat_key` is a good name, and the updated documentation does a great job demonstrating the power that key provides. Awesome.

I think that it might be nice to have a flag that causes the regular repeat key forget capitals, so that Aaron doesn't come out as AAron. By default, the shift state of the stored key would be dropped if Shift was the only modifier and the key was an Alpha 
 (A..Z in English). The function that decides whether a key forgets the shift would probably need to be overridable by the user. This change should probably be proposed in a later PR and not included here, but I wanted to toss it out there to think about.

**Comment by getreuer** [comment]
@Ikcelaks Thank you for that idea! I agree, forgetting Shift could be desirable, especially for doubled letters at the beginning of a capitalized word. I'm happy to add it. I'll split it off to a following PR as you suggest.

**Review by drashna — APPROVED** [review]

**Comment by getreuer** [comment]
Thanks, Drashna!

File: quantum/process_keycode/process_repeat_key.c
Comment (casuanoob) [inline quantum/process_keycode/process_repeat_key.c:46]:
```suggestion
// Default implementation of get_repeat_key_eligible().
__attribute__((weak)) bool get_repeat_key_eligible(uint16_t keycode, keyrecord_t* record) {
    switch (keycode) {
        // Ignore MO, TO, TG, TT, and TL layer switch keys.
        case QK_TRI_LAYER_LOWER:
        case QK_TRI_LAYER_UPPER:
        case QK_MOMENTARY ... QK_MOMENTARY_MAX:
```

@getreuer I'm currently using the userspace implementation of your Repeat key, not the PR, but I feel like the Tri Layer specific layer keycodes should also be included in the default `get_repeat_key_eligible`, since they're essentially MO keycodes with additional function?

**Comment by Ikcelaks** [comment]
I think that there should be a `get_repeat_key_eligible_user` function that is called by the default `get_repeat_key_eligible` function (in the same way that `get_alt_repeat_key_keycode` calls `get_alt_repeat_key_keycode_user`).

Use case: In the past week, I have added my own extra custom keycodes that act as additional alternate repeat keys. They are easy to implement by calling the `get_repeat_keycode` and `get_repeat_key_mods` functions, but it requires overloading the default `get_repeat_key_eligible` to exclude those custom keys.

**Comment by Ikcelaks** [comment]
Here is a link to the `keymap.c` I use, which probably better explains what I meant when I said that I created custom keycodes to act as additional alternate repeat keys. https://github.com/Ikcelaks/keyboard_layouts/blob/main/magic_sturdy/QMK_Layout/Moonlander/magic_sturdy/keymap.c

I don't think that it's necessary to bake in the ability to use multiple alternate repeat keys, because it's almost just as easy to create them yourself as I have, but overriding the `get_repeat_key_eligible` is definitely fragile compared to defining a user function that gets hooked in.

File: quantum/process_keycode/process_repeat_key.c
Comment (getreuer) [reply quantum/process_keycode/process_repeat_key.c]: Great catch! Thanks for the edit. Committed.

**Comment by getreuer** [comment]
@Ikcelaks you make a good point. It isn't useful to repeat modifier and layer switch keys, so these might as well be base logic that is added to by a "`get_repeat_key_eligible_user()`" callback, then users have less to write to ignore an additional key. I just pushed a commit to refactor in this way. (I'll update my userspace implementation this way, too.)

> Here is a link to the keymap.c I use, which probably better explains what I meant when I said that I created custom keycodes to act as additional alternate repeat keys. https://github.com/Ikcelaks/keyboard_layouts/blob/main/magic_sturdy/QMK_Layout/Moonlander/magic_sturdy/keymap.c

Thanks for sharing this! The idea of additional alternate repeat keys is very cool.

**Comment by getreuer** [comment]
@Ikcelaks it is really cool how you did that to make two additional "magic" buttons, and all the mileage you're getting out of them like shortcuts for frequent words and n-grams. I just pushed a commit adding further examples to the docs to illustrate these tricks. After your previous comment ('think of it as the "magic" key'), I've been trying some shortcuts myself and am digging it (got my username on G + AltRepeat = "`getreuer`").

(Off topic) In that [keymap you linked](https://github.com/Ikcelaks/keyboard_layouts/blob/main/magic_sturdy/QMK_Layout/Moonlander/magic_sturdy/keymap.c), what is your layout? 

```
v m l c p     b * u o ,
s t r d y q   f n e a i
x k j g w     z h ' ? .
```

It looks like a mod of the Sturdy layout with `j k` in nicer positions for Vim?... I've been looking for stuff like this. Interested in any info you have on it.

**Comment by Ikcelaks** [comment]
@getreuer , I'm glad you're finding the "magic key" conceptualization useful. In terms of theoretical power, it's equivalent to having a one-shot layer key with the layer restricted to macros that must send the key's original layer keycode first. However, it feels so much better than that, because the key press order is reversed (giving you an extra character to recognize the pattern) and every keypress produces one (or more) characters, which eliminates the lag. I wasn't sure how good it would end up feeling when I first started experimenting a little over a month ago, and I couldn't be more pleased.

The additional magic keys tied to the combos are a very recent experiment. I generally dislike combos, but I knew from my previous experiments with "which", "would", "before", and "because" on the original repeat and alt_repeat keys that full words are extremely easy to remember and recognize (suffixes are also easy; mid-word n-grams and sfb eliminations take more focused practice). Since I'm getting a 5 or 6 key press per combo discount in most cases, I am finding them worthwhile and will probably expand usage there in the future.

In principal, we could further generalize the alt repeat key to a full "follower" key that looked back through the stack of previously pressed keys to an arbitrary depth, but I struggle to come up with any truly compelling examples, and I imagine the extra complication would kill the ease of use factor. In practice, more complex sequences are probably best implemented using your auto-correct feature.

The layout is indeed a mod of Sturdy. Technically, it's a mod on STRD1, which is a variant of Sturdy that uses a vowel layout with a `,i.` pinky column. I'm pretty sure that the `J` being in a better position for VIM is purely coincidence; it was moved to allow `,` and `.` to both be on the pinky. Oxey generated both layouts with Oxeylyzer, and I'm not 100% sure on the motivations. At any rate, STRD1 is perfect for the magic key concept, because the magic key (`^`) is replacing `Z` from the original layout, and that's the best place to put the magic key, because `UE`, `EU`, and `OA` all become inward rolls. It's incredibly lucky for me that I was learning STRD1 when I thought of this concept, because I have yet to find another layout that synergizes half as well. Base Sturdy almost works, but the magic key has a lot of interaction with `.`, so it's less ideal.

File: quantum/repeat_key.c
Comment (sigprof) [inline quantum/repeat_key.c:138]: This does not seem to offer a way to make a predefined alternate key do nothing (apart from returning a keycode which is not handled anywhere, but that would still result in calls to `process_record()`).

Maybe it would be better to redefine the return value of `get_alt_repeat_key_keycode_user()` as:
- `KC_NO` — do nothing (any predefined alternate key is not used);
- `KC_TRNS` — use the predefined alternate key if it exists;
- anything else — use the specified keycode.

```suggestion
    // Call the user callback first to give it a chance to override the default
    // alternate key definitions that follow.
    uint16_t alt_keycode = get_alt_repeat_key_keycode_user(keycode, mods);

    if (alt_keycode != KC_TRANSPARENT) {
        return alt_keycode;
    }
```

File: quantum/repeat_key.c
Comment (sigprof) [inline quantum/repeat_key.c:280]: If the above change is accepted, this would need to be adjusted too:
```suggestion
    return KC_TRANSPARENT;
```

**Review by sigprof — CHANGES_REQUESTED** [review]

File: data/constants/keycodes/keycodes_0.0.2_quantum.hjson
Comment (getreuer) [reply data/constants/keycodes/keycodes_0.0.2_quantum.hjson:16]: Thanks for catching that!

File: quantum/repeat_key.c
Comment (getreuer) [reply quantum/repeat_key.c:280]: Committed.

File: data/schemas/keyboard.jsonschema
Comment (getreuer) [reply data/schemas/keyboard.jsonschema:415]: I appreciate those clarifications and the review! Earlier in this PR, @Ikcelaks made a good suggestion that a following PR could add an option to forget capitals from the repeated mods (use case: so that "Aaron" doesn't come out as "AAron"). I made this object with the intention that such an option would eventually go there. I'll go ahead and add a forget capitals option in this PR.

**Review by getreuer — COMMENTED** [review]
@sigprof thanks so much for the review!

**Comment by getreuer** [comment]
> Earlier in this PR, @Ikcelaks made a good suggestion that a following PR could add an option to forget capitals from the repeated mods (use case: so that "Aaron" doesn't come out as "AAron"). I made this object with the intention that such an option would eventually go there. I'll go ahead and add a forget capitals option in this PR.

On further thought, I realize that optional mod filtering could fit in nicely as an arg to `get_repeat_key_eligible_user()`:

```c
bool get_repeat_key_eligible_user(uint16_t keycode, keyrecord_t* record,
                                  uint8_t* remembered_mods) {
    // Forget Shift on letter keys when Shift or AltGr are the only mods.
    switch (keycode) {
        case KC_A ... KC_Z:
            if ((*remembered_mods & ~(MOD_MASK_SHIFT | MOD_BIT(KC_RALT))) == 0) {
                *remembered_mods &= ~MOD_MASK_SHIFT;
            }
            break;
    }

    return true;
}
```

I just pushed https://github.com/qmk/qmk_firmware/pull/19700/commits/86ac71ff67595ed1dc45cea31bab050ac7b85f66 to do this. There is no need for a separate config option with this approach, so I dropped the changes to info_rules.hjson and keyboard.jsonschema as @sigprof recommended.

**Comment by getreuer** [comment]
To follow up, the recent revisions to `get_repeat_key_eligible_user()` and `get_alt_repeat_key_keycode_user()` are now analogously done in my userspace Repeat Key implementation in https://github.com/getreuer/qmk-keymap/commit/cab494f50dfe8116e6b2a45bf366bd5d85efc904.

@Ikcelaks thank you for that context on STRD1 and Magic Sturdy! I found [your Magic Sturdy writeup](https://github.com/Ikcelaks/keyboard_layouts/blob/main/magic_sturdy/magic_sturdy.md), thanks for doing this. You have gained a user! I've been test driving Magic Sturdy since Friday. As luck would have it, I've been experimenting with a Sturdy mod that was only a few keyswaps away, so I'm getting up to speed quickly and excited to begin getting a "taste of the power" with the right index magic key. That "-ion" shortcut is very nice.

**Comment by getreuer** [comment]
I'm a GitHub noob and have a question. I made the changes that @sigprof requested, but the PR status still shows "Changes requested." Should I be clicking something to clear that, or is that up to reviewers to resolve?

Screenshot

In any case, this PR is ready to resume review.

**Comment by getreuer** [comment]
@Ikcelaks thank you so much for your support and offer to help! Testing interaction with other features would absolutely be appreciated. I also welcome your review comments if you have further thoughts on the code.

> I see that this PR is tagged for breaking_change_2023q2. Does this mean that it's scheduled to make a particular release?

I described in my previous comment that the PR status shows "Changes requested". I have made those changes, but I am a GitHub noob and not certain what the process is from here. I think the next step is that @sigprof would at their discretion clear the "Changes requested" state in a re-review. Let me know if I'm missing something!

Besides this, yes, @drashna applied the "breaking_change_2023q2" tag. My understanding is that this is tentative, since changes to core require at least two approvers to merge. So far, drashna has approved and requested review from other core QMK members. Thank you, drashna!

**Comment by drashna** [comment]
> I'm a GitHub noob and have a question. I made the changes that @sigprof requested, but the PR status still shows "Changes requested." Should I be clicking something to clear that, or is that up to reviewers to resolve?

Nope, don't need to do anything. Github ... is a bit odd in some places.  The sidebar with reviews is more accurate, as the "changes requested" part doesn't actually get updated until a new review is added by that reviewer or is manually cleared.  It's a bit odd. 

> I see that this PR is tagged for breaking_change_2023q2. Does this mean that it's scheduled to make a particular release? Is there anything needed from a testing perspective that I can help with?

It's a way for us to filter PRs, and give a bit higher priority to getting them reviewed, so it's more of a soft target for them.

**Comment by getreuer** [comment]
@drashna thanks for clearing that up! Good to hear the PR is on track, and I appreciate that the breaking_change_2023q2 tag improves review priority a bit.

**Comment by Ikcelaks** [comment]
@getreuer can you merge in the dev branch again at your next convenience?

I thought that I found a bad interaction with Combos, but it turns out that my testing instead ran into the regression bug recently fixed by #20669.

Otherwise, I haven't run into any issues.

**Review by tzarc — APPROVED** [review]

**Comment by drashna** [comment]
looks like `repeat_key_combo` unit test is failing.

**Pull request merged**
